### ISSUE TITLE
websocket support in heplify-server

### ISSUE THREAD

**Issue by aqsyonas** [sipcapture/heplify-server#362]
Hi, 
I am trying to add websocket support in heplify-server. Before i could send PR, i would need little help. 

A widget(pcap importer) from homer-app will send HEP in binary format towards heplify-server's websocket. 
So, i am take data from socket and pass it to hep channel.  Code snippet 
```
for {
	c, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		logp.Err("Error in upgrade", err)
		return
	}
	defer c.Close()
	for {
		buf := h.buffer.Get().([]byte)
		_, buf, err = c.ReadMessage()
		if err != nil {
			logp.Err("got error in while reading data from websocket", err)
			break
		}
		h.inputCh <- buf
		atomic.AddUint64(&h.stats.PktCount, 1)
	}
}
```
But i get `slice bounds out of range [:8192] with capacity 1536` at below line. 
https://github.com/sipcapture/heplify-server/blob/master/server/server.go#L168
Removing the slicing of buffer to maxlenght make it work.  

I know it happens because i try to slice it out of range, but same stuff works for data coming out of udp socket.

i tried to check length of data from udp socket it was shorten than `maxPktLen` but slicing worked irrespective of that. 

I tried to compare code from udp listener to my websocket listener but could not spot any oddity that would not make my websocket code work. 

Surely it is something naive that i am missing.

**Comment by negbie** [comment]
If you remove that line you create a memory leak. WebSocket uses HTTP for the init handshake. After that, WebSocket essentially uses raw TCP so better compare it with TCP code I guess.

This part of your code snippet makes no sense:
buf := h.buffer.Get().([]byte)

ReadMessage already returns []byte

**Comment by aqsyonas** [comment]
@negbie thanks for the pointers, i should be checking with tcp listener code. Could we please keep this open for few days in case i had other question. I would surely close this in few days. 
Thanks you very much. !

**Comment by negbie** [comment]
Sure. You can also take a look at github.com/gobwas/ws. I've used it for some projects and it works very well. It has a better control for lower-level net code.

**Comment by aqsyonas** [comment]
Thanks @negbie for the library link. And sorry for coming late here, was dragged by other things. 
I was actually using [https://github.com/gorilla/websocket](url) for the websocket support due to good stars and easy examples. 

I am still unable to make it work, i have checked tcp and tls listener code but it did not give me much hits(might be due to my naiveness in golang) . 
Below is how i tried to re-write my code. 
```
package input

import (
	"encoding/binary"
	"fmt"
	"github.com/negbie/logp"
	"log"
	"net/http"
	"sync/atomic"

	"github.com/gorilla/websocket"
)

var upgrader = websocket.Upgrader{ReadBufferSize:  1024,
	WriteBufferSize: 1024,
	CheckOrigin: func(r *http.Request) bool {
		return true
	},
}

func (h *HEPInput) relayHandler(w http.ResponseWriter, r *http.Request) {
	c, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		logp.Err("Error in upgrade", err)
		return
	}
	defer c.Close()
	for {
		_, buf, err := c.ReadMessage()
		if err != nil {
			logp.Err("Got error in while reading data from websocket", err)
			break
		}
		size := binary.BigEndian.Uint16(buf[4:6])

		fmt.Println("Message length", len(buf))
		fmt.Println("Size of the message", size)
		fmt.Println("Size of truncate message", len(buf[:size]))
		h.inputCh <- buf[:size]
		atomic.AddUint64(&h.stats.PktCount, 1)
	}
}

func (h *HEPInput) serveWS(addr string) {
	http.HandleFunc("/ws", h.relayHandler)
	err := http.ListenAndServe(addr, nil)
	if err != nil {
		log.Fatal(err)
	}
}
```
And it produces below output. 
```
Message length 1331
Size of the message 1331
Size of truncate message 1331
panic: runtime error: slice bounds out of range [:8192] with capacity 3584
```
Can you give me some hints, how to truncate this message to exact length. So, i can pass this to hep channel?

Or what information am I missing above?

Thanks.

**Comment by negbie** [comment]
You miss the point how in Go sync.Pool works. https://golang.org/pkg/sync/

**Comment by aqsyonas** [comment]
@negbie thanks for the link, tried to go through this and other link related to sync.pool. Had some understanding how this work, but still cannot think of a way to fix this in heplify-server. I will go more reading on this and see if i could make it work. 
If you have some additional pointer, it would surely help. 
Thanks for you guidance here. !

**Comment by negbie** [comment]
@aqsyonas I recommended you already https://github.com/gobwas/ws because it has a better control for lower-level net code. You can get a buffer from the pool, pass it to read TCP payload, make sure that you keep in mind that tcp is a streaming protocol and handle the packet length like you have to do for TCP packets, return buffer to the pool-

**Comment by aqsyonas** [comment]
@negbie i should have saved a lot of time had i taken your advice first. So, i have used the above library for websocket server. And below is how my code looks like.. 

```
func (h *HEPInput) serveWS(addr string) {
	ln, err := net.Listen("tcp", add)
	if err != nil {
		log.Fatal(err)
	}
	for {
		conn, err := ln.Accept()
		if err != nil {
			logp.Err("%v", err)
			return
		}
		_, err = ws.Upgrade(conn)
		if err != nil {
			logp.Err("%v", err)
			return
		}

		go func() {
			defer conn.Close()

			for {
				header, err := ws.ReadHeader(conn)
				if err != nil {
					logp.Err("%v", err)
					return
				}

				buf := h.buffer.Get().([]byte)
				payload := buf[:header.Length]
				_, err = io.ReadFull(conn, payload)
				if err != nil {
					logp.Err("%v", err)
					return
				}

				if header.Masked {
					ws.Cipher(payload, header.Mask, 0)
				}
				// Reset the Masked flag, server frames must not be masked as
				// RFC6455 says.
				header.Masked = false

				if header.OpCode == ws.OpClose {
					return
				}
				h.inputCh <- payload
				atomic.AddUint64(&h.stats.PktCount, 1)
			}
		}()
	}
}
```
I see no more crashes in heplify-server and no errors in heplify logs. I see heplify-server parse the pcap just fine. Do you have any objection/recommendation for above snippet?
Many thanks for your help.

**Comment by negbie** [comment]
@aqsyonas that's basicly the way I would have done it. Now you need to make sure to clean up everything if func (h *HEPInput) End() is called.

**Comment by aqsyonas** [comment]
@negbie thank you very much. I have created a pull request for the feature, please feel free to leave comments for improvements or modifications as you seem good

**Comment by aqsyonas** [comment]
Closed in #376

Thanks Again for the help. !

**Comment by negbie** [comment]
Your welcome!
